drivers/sensor/iis2iclx/iis2iclx.c

```diff
@@ -0,0 +1,657 @@
+	val->val1 = rh;
+	val->val2 = rh * 1000000;
+}
+
+static inline void iis2iclx_press_convert(struct sensor_value *val,
+					    struct iis2iclx_data *data)
+{
+	int32_t raw_val;
+	int idx;
+
+	idx = iis2iclx_shub_get_idx(SENSOR_CHAN_PRESS);
+	if (idx < 0) {
+		LOG_DBG("external press/temp not supported");
+		return;
+	}
+
+	raw_val = (int32_t)(data->ext_data[idx][0] |
+			  (data->ext_data[idx][1] << 8) |
+			  (data->ext_data[idx][2] << 16));
+
+	/* Pressure sensitivity is 4096 LSB/hPa */
+	/* Convert raw_val to val in kPa */
+	val->val1 = (raw_val >> 12) / 10;
+	val->val2 = (raw_val >> 12) % 10 * 100000 +
+		(((int32_t)((raw_val) & 0x0FFF) * 100000L) >> 12);
+}
+
+static inline void iis2iclx_temp_convert(struct sensor_value *val,
+					   struct iis2iclx_data *data)
+{
+	int16_t raw_val;
+	int idx;
+
+	idx = iis2iclx_shub_get_idx(SENSOR_CHAN_PRESS);
+	if (idx < 0) {
+		LOG_DBG("external press/temp not supported");
+		return;
+	}
+
+	raw_val = (int16_t)(data->ext_data[idx][3] |
+			  (data->ext_data[idx][4] << 8));
+
+	/* Temperature sensitivity is 100 LSB/deg C */
+	val->val1 = raw_val / 100;
+	val->val2 = (int32_t)raw_val % 100 * (10000);
+}
+#endif
+
+static int iis2iclx_channel_get(const struct device *dev,
+				  enum sensor_channel chan,
+				  struct sensor_value *val)
+{
+	struct iis2iclx_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_ACCEL_XYZ:
+		iis2iclx_accel_channel_get(chan, val, data);
+		break;
+#if defined(CONFIG_IIS2ICLX_ENABLE_TEMP)
+	case SENSOR_CHAN_DIE_TEMP:
+		iis2iclx_temp_channel_get(val, data);
+		break;
+#endif
+#if defined(CONFIG_IIS2ICLX_SENSORHUB)
+	case SENSOR_CHAN_MAGN_X:
+	case SENSOR_CHAN_MAGN_Y:
+	case SENSOR_CHAN_MAGN_Z:
+	case SENSOR_CHAN_MAGN_XYZ:
+		iis2iclx_magn_get_channel(chan, val, data);
+		break;
+
+	case SENSOR_CHAN_HUMIDITY:
+		iis2iclx_hum_convert(val, data);
+		break;
+
+	case SENSOR_CHAN_PRESS:
+		iis2iclx_press_convert(val, data);
+		break;
+
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		iis2iclx_temp_convert(val, data);
+		break;
+#endif
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static const struct sensor_driver_api iis2iclx_api_funcs = {
+	.attr_set = iis2iclx_attr_set,
+#if CONFIG_IIS2ICLX_TRIGGER
+	.trigger_set = iis2iclx_trigger_set,
+#endif
+	.sample_fetch = iis2iclx_sample_fetch,
+	.channel_get = iis2iclx_channel_get,
+};
+
+static int iis2iclx_init_chip(const struct device *dev)
+{
+	struct iis2iclx_data *iis2iclx = dev->data;
+	uint8_t chip_id;
+
+	iis2iclx->dev = dev;
+
+	if (iis2iclx_device_id_get(iis2iclx->ctx, &chip_id) < 0) {
+		LOG_DBG("Failed reading chip id");
+		return -EIO;
+	}
+
+	LOG_INF("chip id 0x%x", chip_id);
+
+	if (chip_id != IIS2ICLX_ID) {
+		LOG_DBG("Invalid chip id 0x%x", chip_id);
+		return -EIO;
+	}
+
+	/* reset device */
+	if (iis2iclx_reset_set(iis2iclx->ctx, 1) < 0) {
+		return -EIO;
+	}
+
+	k_busy_wait(100);
```

Comment:
Same question as before, can you sleep instead?

Reply:
I can use k_sleep